<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Joyce Hsiao" />

<meta name="date" content="2019-05-07" />

<title>Effect of data transformation in null data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">scDE</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/dsc-log-fold-change">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Effect of data transformation in null data</h1>
<h4 class="author"><em>Joyce Hsiao</em></h4>
<h4 class="date"><em>2019-05-07</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-05-10
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>dsc-log-fold-change/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.3.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181115code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181115)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181115code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181115)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabdsclogfoldchangetree4f25ca2abdb1357f3135519359f85c31f7c3fd6ftargetblank4f25ca2a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/stephenslab/dsc-log-fold-change/tree/4f25ca2abdb1357f3135519359f85c31f7c3fd6f" target="_blank">4f25ca2</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabdsclogfoldchangetree4f25ca2abdb1357f3135519359f85c31f7c3fd6ftargetblank4f25ca2a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .sos/
    Ignored:    analysis/.sos/
    Ignored:    dsc/.sos/
    Ignored:    dsc/benchmark/
    Ignored:    dsc/data/filtered_gene_bc_matrices/
    Ignored:    dsc/dsc_test/.sos/
    Ignored:    output/

Unstaged changes:
    Modified:   analysis/eval_initial.Rmd
    Modified:   dsc/benchmark.dsc
    Modified:   dsc/benchmark.sh
    Modified:   dsc/modules/transform_data.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/4f25ca2abdb1357f3135519359f85c31f7c3fd6f/analysis/transform_null.Rmd" target="_blank">4f25ca2</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
<td>
update results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/dsc-log-fold-change/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/transform_null.html" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/5d43eb7674e12a4f95b9e355bd26a2c0139930fe/analysis/transform_null.Rmd" target="_blank">5d43eb7</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
<td>
update results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/dsc-log-fold-change/5c509febc60baae0a0e99bca5a138e198ba71586/docs/transform_null.html" target="_blank">5c509fe</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/9881ccac37a5df70a3dd97bb3840a6c981c68c9c/analysis/transform_null.Rmd" target="_blank">9881cca</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-09
</td>
<td>
add summary of the results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/dsc-log-fold-change/fcfa1efd7fd093268e7276e8e9fba25c3df1fe34/docs/transform_null.html" target="_blank">fcfa1ef</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/70a66d1b238b0e00aa2ae8445bdf5fe529e8607f/analysis/transform_null.Rmd" target="_blank">70a66d1</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
update results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/dsc-log-fold-change/7489eadba2d2171d86d984d163db74ff3ac266fc/docs/transform_null.html" target="_blank">7489ead</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/dd437668279190aad1f8fe8e00c64055b9009668/analysis/transform_null.Rmd" target="_blank">dd43766</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
update results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/dsc-log-fold-change/a4c55012d3334d1c248052982c60151cb14306ec/docs/transform_null.html" target="_blank">a4c5501</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/eafe1738c1fc80586d60d4daa81cfe60e1044f3b/analysis/transform_null.Rmd" target="_blank">eafe173</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-08
</td>
<td>
downstream effects of data transformation methods in null datasets: initial assessment
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>Applying data transformation methods to single-cell gene expression count and computing type I error rate in null datasets.</p>
<ul>
<li><p><a href="#no-differences-in-library-sizes">When library sizes are the same between two groups</a>, except for edgeR which has poor control over type I error at alpha &lt; .01, all other methods are able to control for type I error.</p></li>
<li><p><a href="#library-size-factor-1">When sample library size in one group is 2 times the library size in the other group</a> all methods perform pooly, except for the method that applies t-test to Pearson’s residuals (Hafemeister and Satija, 2019). We observed <a href="#library-size-factor-2">similar results after increasing library size factor to 2</a> (group 1 sample size <span class="math inline">\(2^2\)</span> times of group 2 sample size).</p></li>
</ul>
<hr />
<p>Data transformation methods compared:</p>
<p><span class="math inline">\(X_{ig}\)</span>: gene expression count in cell <span class="math inline">\(i\)</span> from gene <span class="math inline">\(g\)</span></p>
<p><span class="math inline">\(X_{i+}\)</span>: gene expression count in cell <span class="math inline">\(i\)</span>, i.e., <span class="math inline">\(X_{i+} = \sum_g X_{ig}\)</span></p>
<p><span class="math inline">\(S_i\)</span>: library size normalization factor for cell <span class="math inline">\(i\)</span></p>
<p><span class="math inline">\(p\)</span>: a positive pseudo-count added to expression matrix; traditionally it is used to ensure log-transformation of the expression matrix is well-defined. For now we use pseudo-count of 1.</p>
<ol style="list-style-type: decimal">
<li><p><code>log2_none_p1</code>: <span class="math inline">\(log2(X_{ig} + p)\)</span> where <span class="math inline">\(p=1\)</span>.</p></li>
<li><p><code>log2_libsum_p1</code>: <span class="math inline">\(log2(X_{ig}/X_{i+} + p)\)</span> where <span class="math inline">\(p=1\)</span>. Commonly know as <code>total sum normalization</code> or <code>tota count normalization</code>. This methods assumes that the true library size is the same across cells.</p></li>
<li><p><code>log2_libscale_TMM_p1</code>: <span class="math inline">\(log2(X_{ig}/S_{i} + p)\)</span> where <span class="math inline">\(p=1\)</span> using edgeR TMM method to estimate <span class="math inline">\(S_i\)</span>.</p></li>
<li><p><code>log2_libscale_RLE_p1</code>: <span class="math inline">\(log2(X_{ig}/S_{i} + p)\)</span> where <span class="math inline">\(p=1\)</span> using DESeq2 RLE (Relative Log Expression) method to estimate <span class="math inline">\(S_i\)</span>. This method assumes that gene expression distribution across samples should be similar, and to account for outlier genes, they let the medians of gene expression distribution to be the same across samples (cells). The size factor of the library size is estimated to be the median of the ratios of observed counts:</p></li>
</ol>
<p><span class="math display">\[
\widehat{S}_i = \text{median}_g \frac{X_{ig}}{(\prod^N_{n=1}X_{ig})^{1/N}}
\]</span></p>
<ol start="5" style="list-style-type: decimal">
<li><code>counts_pearsons</code>: Pearson’s residuals of expression counts, derived using <code>sctransform</code> (Hafemeister and Satija, 2019).</li>
</ol>
<hr />
<p>Pipelines compared:</p>
<ul>
<li>DEseq2: count data + RLE transformation</li>
<li>edgeR: count data + TMM transformation</li>
<li>limma_voom + data transformation 1-4</li>
<li>t-test + all 5 data transformation methods listed above</li>
</ul>
<hr />
<p>More about <code>counts_pearson</code>:</p>
<p>For a given gene <span class="math inline">\(g\)</span>, use the sum of all molecules assigned to a cell as a proxy for sequencing depth, and use this cell attribute in a regression model with negative binomial distribution and log link function. Thus, let <span class="math inline">\(X_g\)</span> be the vector of UMI counts assigned to gene <span class="math inline">\(g\)</span>, and <span class="math inline">\(m\)</span> be the vector of molecules assigned to the cells, i.e., <span class="math inline">\(m_i = \sum_i X_{ig}\)</span>. For a given <span class="math inline">\(g\)</span>, we have</p>
<p><span class="math inline">\(log(E(X_g)) = \beta_0 + \beta_1 log10 m\)</span></p>
<p>Using the NB parametrization with mean <span class="math inline">\(\mu\)</span> and variance <span class="math inline">\(\mu + \mu^2/\theta\)</span>,</p>
<p>Pearson’s residuals are defined as:</p>
<p><span class="math inline">\(z_{ig} = (X_{ig}-\mu_{ig})/\sigma_{ig}\)</span></p>
<p>where</p>
<p><span class="math inline">\(\mu_{ig} = exp(\beta_{0g} + \beta_{1g}log10 m_i)\)</span>,</p>
<p><span class="math inline">\(\sigma_{ig} = \sqrt(\mu_{ig} + \mu^2_{ig}/\theta_{g})\)</span></p>
<hr />
<p>Data simulation parameters:</p>
<ul>
<li>100 cells (50 vs 50)</li>
<li>libsize_factor: 0, 1, 2 (group 1 is 2^0, 2^1 and 2^2 of group 2)</li>
<li>1K genes randomly drawn</li>
<li>100% null genes</li>
<li>50 simulated datasets</li>
</ul>
<hr />
</div>
<div id="required-packages-and-functions" class="section level2">
<h2>Required packages and functions</h2>
<pre class="r"><code>library(tidyverse)
library(seqgendiff)
library(sctransform)
#source(&quot;dsc/modules/poisthin.R&quot;)
source(&quot;dsc/modules/filter_genes.R&quot;)
source(&quot;dsc/modules/transform_data.R&quot;)
source(&quot;dsc/modules/t_test.R&quot;)
source(&quot;dsc/modules/wilcoxon.R&quot;)
source(&quot;dsc/modules/limma_voom.R&quot;)
source(&quot;dsc/modules/edger.R&quot;)
source(&quot;dsc/modules/deseq2.R&quot;)
source(&quot;dsc/modules/sva.R&quot;)

run_methods &lt;- function(data_obj, nsim, verbose=F) {
  
  Y &lt;- t(data_obj$Y)
  X &lt;- data_obj$X
  beta &lt;- data_obj$beta
  keep_genes &lt;- filter_genes(Y, min_cell_detected=5)
  Y &lt;- Y[keep_genes,]
  beta &lt;- beta[keep_genes]
  
  foo_m &lt;- do.call(rbind, lapply(1:length(de_methods_list), function(j) {

  if (de_methods_list[j] == &quot;edger&quot;) {
    if (verbose) message(&quot;edger&quot;)
      res &lt;- edger(Y=Y, X=X)
      pvals &lt;- res$pval
      return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                        type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                        mse = mean((res$est - beta)^2, na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == &quot;deseq2&quot;) {
      if (verbose) message(&quot;deseq2&quot;)
      res &lt;- deseq2(Y=Y, X=X)
      pvals &lt;- res$pval
      return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                        type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                        mse = mean((res$est - beta)^2, na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == &quot;limma&quot;) {
      if (verbose) message(&quot;limma&quot;)
      foo_l &lt;- do.call(rbind, lapply(1:(length(transform_methods_list)-1), function(k) {
          if (transform_methods_list[k] == &quot;log2_none_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;none&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libsum_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;sum&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libscale_TMM_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;TMM&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libscale_RLE_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;RLE&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (sum(is.na(transformed_Y))==nrow(transformed_Y)*ncol(transformed_Y)) {
          return(data.frame(type1error_01=NA,
                            type1error_001=NA,
                            mse=NA,
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
          } else {
          fit &lt;- lmFit(transformed_Y,design=X)
          fit.ebayes &lt;- eBayes(fit)
          pvals &lt;- fit.ebayes$p.value[,2]
          return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                            type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                            mse=mean((fit.ebayes$coefficients[,2]-beta)^2,na.rm=T),
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
          }
      }) )
      return(foo_l)      
    }
  if (de_methods_list[j] == &quot;limma_voom&quot;) {
      if (verbose) message(&quot;limma_voom&quot;)
      res &lt;- limma_voom(Y=Y, X=X)
      pvals &lt;- res$pvalue
      return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                        type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                        mse=mean((res$betahat-beta)^2,na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == &quot;limma_voom_libscale_tmm&quot;) {
  if (verbose) message(&quot;limma_voom_libscale_tmm&quot;)
  libnorm_factors &lt;- edgeR::calcNormFactors(Y, method=&quot;TMM&quot;)
  res &lt;- limma_voom(Y, X, libnorm_factors=libnorm_factors)
  pvals &lt;- res$pvalue
  return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                    type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                    mse=mean((res$betahat-beta)^2,na.rm=T),
                    transform_method = de_methods_list[j],
                    de_method = de_methods_list[j],
                    nsim = nsim))
  } 
  if (de_methods_list[j] == &quot;limma_voom_libscale_rle&quot;) {
    if (verbose) message(&quot;limma_voom_libscale_rle&quot;)
    libnorm_factors &lt;- edgeR::calcNormFactors(Y, method=&quot;RLE&quot;)
    if (anyNA(libnorm_factors)) {
    return(data.frame(type1error_01=NA,
                      type1error_001=NA,
                      mse=NA,
                      transform_method = de_methods_list[j],
                      de_method = de_methods_list[j],
                      nsim = nsim))
    } else {
    res &lt;- limma_voom(Y, X, libnorm_factors=libnorm_factors)
    pvals &lt;- res$pvalue
    return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                      type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                      mse=mean((res$betahat-beta)^2,na.rm=T),
                      transform_method = de_methods_list[j],
                      de_method = de_methods_list[j],
                      nsim = nsim))
    }
  } 
  if (de_methods_list[j] == &quot;sva_limma_voom&quot;) {
      if (verbose) message(&quot;sva_limma_voom&quot;)
      output_sva &lt;- sva(Y, X)
      res &lt;- limma_voom(Y, X=output_sva$X.sv)
      pvals &lt;- res$pval
      return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                        type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                        mse=mean((res$betahat-beta)^2,na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == &quot;t_test&quot;) {
      if (verbose) message(&quot;t_test&quot;)
      foo_t &lt;- do.call(rbind, lapply(1:length(transform_methods_list), function(k) {
          if (transform_methods_list[k] == &quot;log2_none_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;none&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libsum_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;sum&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libscale_TMM_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;TMM&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;log2_libscale_RLE_p1&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;RLE&quot;, 
                                          log=&quot;log2&quot;, pseudo_count=1)
          }
          if (transform_methods_list[k] == &quot;counts_pearsons&quot;) {
          transformed_Y &lt;- transform_data(Y, libscale_method = &quot;pearsons_residual&quot;, 
                                          log=&quot;none&quot;, pseudo_count=1)
          }
          res &lt;- t_test(transformed_Y, X)
          pvals &lt;- res[2,]
          return(data.frame(type1error_01=mean(pvals &lt; .01, na.rm=TRUE),
                            type1error_001=mean(pvals &lt; .001, na.rm=TRUE),
                            mse=mean((res[1,]-beta)^2,na.rm=T),
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
      }) )
      return(foo_t)      
    }
  }) )
   return(foo_m)
}


transform_methods_list &lt;- c(&quot;log2_none_p1&quot;, &quot;log2_libsum_p1&quot;, &quot;log2_libscale_TMM_p1&quot;, 
                            &quot;log2_libscale_RLE_p1&quot;, &quot;counts_pearsons&quot;)
de_methods_list &lt;- c(&quot;edger&quot;, &quot;deseq2&quot;, &quot;limma&quot;, &quot;limma_voom&quot;, 
                     &quot;limma_voom_libscale_tmm&quot;,
                     &quot;limma_voom_libscale_rle&quot;,
                     &quot;sva_limma_voom&quot;, &quot;t_test&quot;)</code></pre>
<hr />
</div>
<div id="no-differences-in-library-sizes" class="section level2">
<h2>No differences in library sizes</h2>
<p>Simulations</p>
<pre class="r"><code>counts &lt;- readRDS(&quot;dsc/data/pbmc_counts.rds&quot;)
nsamp &lt;- 100
ngene &lt;- 1000
prop_null &lt;- 0
libsize_factor &lt;- 0
signal_fun &lt;- function(n) rep(libsize_factor, n)
signal_params &lt;- list()
#pvals_thres &lt;- .001
nsim &lt;- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj &lt;- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0(&quot;output/transform_null.Rmd/data_obj_&quot;,i,&quot;.rds&quot;))
}



out &lt;- do.call(rbind, lapply(1:nsim, function(i) {
  print(i)
  data_obj &lt;- readRDS(file = paste0(&quot;output/transform_null.Rmd/data_obj_&quot;,i,&quot;.rds&quot;))

  res &lt;- run_methods(data_obj, i, verbose=T)
  return(res)
  }) )

out$method &lt;- as.character(out$de_method)
which_relabel &lt;- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] &lt;- paste(as.character(out$de_method), as.character(out$transform_method), sep=&quot;_&quot;)[which_relabel]

out$method &lt;- factor(out$method)

saveRDS(out, file = &quot;output/transform_null.Rmd/libfactor_0.rds&quot;)</code></pre>
<p>Results</p>
<pre class="r"><code>#alpha &lt;- .001
out &lt;- readRDS(file = &quot;output/transform_null.Rmd/libfactor_0.rds&quot;)
levels(out$method)</code></pre>
<pre><code> [1] &quot;deseq2&quot;                      &quot;edger&quot;                      
 [3] &quot;limma_log2_libscale_RLE_p1&quot;  &quot;limma_log2_libscale_TMM_p1&quot; 
 [5] &quot;limma_log2_libsum_p1&quot;        &quot;limma_log2_none_p1&quot;         
 [7] &quot;limma_voom&quot;                  &quot;limma_voom_libscale_rle&quot;    
 [9] &quot;limma_voom_libscale_tmm&quot;     &quot;sva_limma_voom&quot;             
[11] &quot;t_test_counts_pearsons&quot;      &quot;t_test_log2_libscale_RLE_p1&quot;
[13] &quot;t_test_log2_libscale_TMM_p1&quot; &quot;t_test_log2_libsum_p1&quot;      
[15] &quot;t_test_log2_none_p1&quot;        </code></pre>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .001&quot;) +
        geom_hline(yintercept = .001, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results-1">
Past versions of results-1.png
</button>
</p>
<div id="fig-results-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results-1.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .01&quot;) +
        geom_hline(yintercept = .01, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results-2">
Past versions of results-2.png
</button>
</p>
<div id="fig-results-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results-2.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(nsim, method) %&gt;%
    summarise(mse_mn =mean(mse, na.rm=T)) %&gt;%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Mean squared error&quot;) +
        ylab(&quot;Mean squared error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results-3.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results-3">
Past versions of results-3.png
</button>
</p>
<div id="fig-results-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results-3.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr />
</div>
<div id="library-size-factor-1" class="section level2">
<h2>library size factor 1</h2>
<p>Simulation</p>
<pre class="r"><code>counts &lt;- readRDS(&quot;dsc/data/pbmc_counts.rds&quot;)
nsamp &lt;- 100
ngene &lt;- 1000
prop_null &lt;- 0
libsize_factor &lt;- 1
signal_fun &lt;- function(n) rep(libsize_factor, n)
signal_params &lt;- list()
#pvals_thres &lt;- .001
nsim &lt;- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj &lt;- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0(&quot;output/transform_null.Rmd/data_obj_libfactor_1_&quot;,i,&quot;.rds&quot;))
}


nsim=50
out &lt;- do.call(rbind, lapply(1:nsim, function(i) {

  data_obj &lt;- readRDS(file = paste0(&quot;output/transform_null.Rmd/data_obj_libfactor_1_&quot;,
                                    i,&quot;.rds&quot;))

  res &lt;- run_methods(data_obj, i)
  return(res)
  }) )

out$method &lt;- as.character(out$de_method)
which_relabel &lt;- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] &lt;- paste(as.character(out$de_method), as.character(out$transform_method), sep=&quot;_&quot;)[which_relabel]
out$method &lt;- factor(out$method)

saveRDS(out, file = &quot;output/transform_null.Rmd/libfactor_1.rds&quot;)</code></pre>
<p>Results</p>
<pre class="r"><code>#alpha &lt;- .001
out &lt;- readRDS(file = &quot;output/transform_null.Rmd/libfactor_1.rds&quot;)

levels(out$method)</code></pre>
<pre><code> [1] &quot;deseq2&quot;                      &quot;edger&quot;                      
 [3] &quot;limma_log2_libscale_RLE_p1&quot;  &quot;limma_log2_libscale_TMM_p1&quot; 
 [5] &quot;limma_log2_libsum_p1&quot;        &quot;limma_log2_none_p1&quot;         
 [7] &quot;limma_voom&quot;                  &quot;limma_voom_libscale_rle&quot;    
 [9] &quot;limma_voom_libscale_tmm&quot;     &quot;sva_limma_voom&quot;             
[11] &quot;t_test_counts_pearsons&quot;      &quot;t_test_log2_libscale_RLE_p1&quot;
[13] &quot;t_test_log2_libscale_TMM_p1&quot; &quot;t_test_log2_libsum_p1&quot;      
[15] &quot;t_test_log2_none_p1&quot;        </code></pre>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .001&quot;) +
        geom_hline(yintercept = .001, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_1-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_1-1">
Past versions of results_libfactor_1-1.png
</button>
</p>
<div id="fig-results_libfactor_1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_1-1.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .01&quot;) +
        geom_hline(yintercept = .01, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_1-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_1-2">
Past versions of results_libfactor_1-2.png
</button>
</p>
<div id="fig-results_libfactor_1-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_1-2.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(nsim, method) %&gt;%
    summarise(mse_mn =mean(mse, na.rm=T)) %&gt;%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Mean squared error&quot;) +
        ylab(&quot;Mean squared error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_1-3.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_1-3">
Past versions of results_libfactor_1-3.png
</button>
</p>
<div id="fig-results_libfactor_1-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_1-3.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr />
</div>
<div id="library-size-factor-2" class="section level2">
<h2>library size factor 2</h2>
<pre class="r"><code>counts &lt;- readRDS(&quot;dsc/data/pbmc_counts.rds&quot;)
nsamp &lt;- 100
ngene &lt;- 1000
prop_null &lt;- 0
libsize_factor &lt;- 2
signal_fun &lt;- function(n) rep(libsize_factor, n)
signal_params &lt;- list()
#pvals_thres &lt;- .001
nsim &lt;- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj &lt;- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0(&quot;output/transform_null.Rmd/data_obj_libfactor_2_&quot;,i,&quot;.rds&quot;))
}


nsim=50
out &lt;- do.call(rbind, lapply(1:nsim, function(i) {
  print(i)
  data_obj &lt;- readRDS(file = paste0(&quot;output/transform_null.Rmd/data_obj_libfactor_2_&quot;,
                                    i,&quot;.rds&quot;))

  res &lt;- run_methods(data_obj, i, verbose = T)
  return(res)
  }) )
out$method &lt;- as.character(out$de_method)
which_relabel &lt;- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] &lt;- paste(as.character(out$de_method), as.character(out$transform_method), sep=&quot;_&quot;)[which_relabel]
out$method &lt;- factor(out$method)

saveRDS(out, file = &quot;output/transform_null.Rmd/libfactor_2.rds&quot;)</code></pre>
<p>Results</p>
<pre class="r"><code>#alpha &lt;- .001
out &lt;- readRDS(file = &quot;output/transform_null.Rmd/libfactor_2.rds&quot;)

levels(out$method)</code></pre>
<pre><code> [1] &quot;deseq2&quot;                      &quot;edger&quot;                      
 [3] &quot;limma_log2_libscale_RLE_p1&quot;  &quot;limma_log2_libscale_TMM_p1&quot; 
 [5] &quot;limma_log2_libsum_p1&quot;        &quot;limma_log2_none_p1&quot;         
 [7] &quot;limma_voom&quot;                  &quot;limma_voom_libscale_rle&quot;    
 [9] &quot;limma_voom_libscale_tmm&quot;     &quot;sva_limma_voom&quot;             
[11] &quot;t_test_counts_pearsons&quot;      &quot;t_test_log2_libscale_RLE_p1&quot;
[13] &quot;t_test_log2_libscale_TMM_p1&quot; &quot;t_test_log2_libsum_p1&quot;      
[15] &quot;t_test_log2_none_p1&quot;        </code></pre>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .001&quot;) +
        geom_hline(yintercept = .001, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_boxplot).</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_summary).</code></pre>
<pre><code>Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_2-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_2-1">
Past versions of results_libfactor_2-1.png
</button>
</p>
<div id="fig-results_libfactor_2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_2-1.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(method) %&gt;%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .01&quot;) +
        geom_hline(yintercept = .01, col=&quot;gray50&quot;) +
        ylab(&quot;Type I error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_boxplot).</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_summary).</code></pre>
<pre><code>Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_2-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_2-2">
Past versions of results_libfactor_2-2.png
</button>
</p>
<div id="fig-results_libfactor_2-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_2-2.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>out %&gt;% #filter(n1==50) %&gt;% 
    group_by(nsim, method) %&gt;%
    summarise(mse_mn =mean(mse, na.rm=T)) %&gt;%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab(&quot;Mean squared error&quot;) +
        ylab(&quot;Mean squared error&quot;) +
      scale_x_discrete(position = &quot;top&quot;) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_boxplot).</code></pre>
<pre><code>Warning: Removed 3 rows containing non-finite values (stat_summary).</code></pre>
<pre><code>Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<p><img src="figure/transform_null.Rmd/results_libfactor_2-3.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_libfactor_2-3">
Past versions of results_libfactor_2-3.png
</button>
</p>
<div id="fig-results_libfactor_2-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/dsc-log-fold-change/blob/e6fb98b6639f547c90e4d6c2fc4586fbd48fab2f/docs/figure/transform_null.Rmd/results_libfactor_2-3.png" target="_blank">e6fb98b</a>
</td>
<td>
Joyce Hsiao
</td>
<td>
2019-05-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr />
<p>TBD</p>
<pre class="r"><code>res_libfactor_null &lt;- readRDS(file = &quot;output/transform_null.Rmd/type1error.rds&quot;)
res_libfactor_null$libfactor &lt;- 2^0
res_libfactor_1 &lt;- readRDS(file = &quot;output/transform_null.Rmd/type1error_libfactor_1.rds&quot;)
res_libfactor_1$libfactor &lt;- 2^1
res_libfactor_2 &lt;- readRDS(file = &quot;output/transform_null.Rmd/type1error_libfactor_2.rds&quot;)
res_libfactor_2$libfactor &lt;- 2^2

res &lt;- rbind(res_libfactor_null,
             res_libfactor_1)
res$libfactor &lt;- factor(res$libfactor)


# res %&gt;% #filter(n1==50) %&gt;% 
#     group_by(libfactor, transform_method) %&gt;%
#     ggplot(., aes(x=transform_method, y=type1error_001, col=transform_method)) +
# #        facet_wrap(~libfactor) +
#         geom_boxplot() + geom_point() + xlab(&quot;Type 1 error at alpha &lt; .001&quot;) +
#         geom_hline(yintercept = .001, col=&quot;gray50&quot;) +
#         ylab(&quot;Type I error&quot;) +
#       scale_x_discrete(position = &quot;top&quot;,
#                        labels=labels_methods) +
#        theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
#        stat_summary(fun.y=mean, geom=&quot;point&quot;, shape=4, size=4, col=&quot;black&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sctransform_0.2.0 seqgendiff_0.1.0  forcats_0.3.0    
 [4] stringr_1.3.1     dplyr_0.8.0.1     purrr_0.3.2      
 [7] readr_1.3.1       tidyr_0.8.3       tibble_2.1.1     
[10] ggplot2_3.1.0     tidyverse_1.2.1  

loaded via a namespace (and not attached):
 [1] tidyselect_0.2.5   listenv_0.7.0      reshape2_1.4.3    
 [4] haven_1.1.2        lattice_0.20-38    colorspace_1.3-2  
 [7] generics_0.0.2     htmltools_0.3.6    yaml_2.2.0        
[10] rlang_0.3.4        pillar_1.3.1       glue_1.3.0        
[13] withr_2.1.2        modelr_0.1.2       readxl_1.1.0      
[16] plyr_1.8.4         munsell_0.5.0      gtable_0.2.0      
[19] workflowr_1.3.0    cellranger_1.1.0   rvest_0.3.2       
[22] future_1.12.0      codetools_0.2-15   evaluate_0.12     
[25] labeling_0.3       knitr_1.20         parallel_3.5.1    
[28] broom_0.5.1        Rcpp_1.0.1         scales_1.0.0      
[31] backports_1.1.2    jsonlite_1.6       fs_1.2.6          
[34] gridExtra_2.3      hms_0.4.2          digest_0.6.18     
[37] stringi_1.2.4      grid_3.5.1         rprojroot_1.3-2   
[40] cli_1.0.1          tools_3.5.1        magrittr_1.5      
[43] lazyeval_0.2.1     crayon_1.3.4       whisker_0.3-2     
[46] future.apply_1.2.0 pkgconfig_2.0.2    MASS_7.3-51.1     
[49] Matrix_1.2-17      xml2_1.2.0         lubridate_1.7.4   
[52] assertthat_0.2.0   rmarkdown_1.10     httr_1.3.1        
[55] rstudioapi_0.10    globals_0.12.4     R6_2.4.0          
[58] nlme_3.1-137       git2r_0.23.0       compiler_3.5.1    </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
