---
title: "Home"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
---

Project website for benchmarking methods for single-cell DE methods. 


## Analysis

### Sample size 

* Questions
    * Performances undering large sample size (e.g., 50/group, 250/group)
    * Focusing on generic methods: not yet tailored methods for single-cell data
    * Do bulk methods perform as expected in bulk data?

* Simulations based on PBMC data
    * [Effect size, power and sample size](eval_initial.html)
    * [Power and false discovery rate](eval_initial_power.html)
    * [Type I error](eval_initial_type1.html)    
    
* Simulations based on GTEx lung data
    * [Type I error](gtex_type1.html)    
    * [Power and false discovery rate](gtex_power.html)

## Data transformation

* [In null datasets, assess the ability of methods to control type I error, including edger, deseq2, limma + voom, t-test + simple data transformation, and wilcoxon.](transform_null.html)



### Data we used 

* [README](dsc/data/README.md)
* 10x Genomics dataset: ~3K PBMC samples
* GTEx lung samples

---

## One-time investigations

* Replicate results in Van den Berge et al., 2018
    * Curate PBMC dataset using [their code](pbmc_berge_null.html)
    * Simulate scRNA-seq data with signals using [their code](sim_power_berge_pbmc.html)
    
    


    

