---
title: "Evaluate two-sample tests"
author: "jhsiao999"
date: "2018-11-15"
output: workflowr::wflow_html
---

## Overview

We applied two-sample tests for mean difference to simulated datasets.

The methods applied include:

* DESeq2
* glm poisson
* glm quasipoisson
* t-test
* wilcoson tet

The simulated data were made from experimental datasets as described below. 

1. Draw (sample without replacement) p genes and N samples from the experimental dataset. 

2. For the `Correlated case`, assign samples to Group 1 or Group 2. 

3. For the `Uncorrelated case`, permute sample labels for each gene, then in the permuted sample, assign samples to Group 1 or Group 2. 

In addition, the parameter setting is:

* Sample size (n1 vs n2): 50 vs 50, 300 vs 300

* Number of genes (p): 946 

Regarding the experimental data:

* single cell RNA-seq data from 10x genomics technology

* The dataset used in the simulation include CD14+ Monocytes and CD8 T cells, a total of 787 samples and 946 genes. 

* The original complete dataset include 2,700 single cells and 32,738 genes from 8 immune cell types: B cells, CD4 T cells, CD8 T cells, CD14+ Monocytes, Dendritic cells, FCGR3A+ Monocytes, Megakaryocytes, NK cells. 

* We filtered the original data to include samples that are detected in > 200 genes and genes that are detected in > 20% of cells.


---

## Analysis

Running job on midway2

```{bash, eval=F}
cd dsc

dsc ./benchmark.dsc --host ./code/config.yaml -c 20
```

```{r, warning=F, message=F}
get_res <- function(dir_dsc, dscoutput, 
                     sim_case=c("sample_correlated", "sample_uncorrelated"), 
                     sample_size=c(50,300)) {
 
  n_methods <- nlevels(dscoutput$method)
  out.sub <- dscoutput[dscoutput$get_data==sim_case & dscoutput$get_data.n1==sample_size,]
  
  res <- vector("list",n_methods)
  for (i in 1:nrow(out.sub)) {
#    print(i)
    fl <- readRDS(file.path(dir_dsc,
                         paste0(as.character(out.sub$method.p[i]), ".rds")))
    res[[i]] <- data.frame(method = as.character(out.sub$method)[i],
                           seed = out.sub$get_data.seed[i],
                           n1_n2 = paste0(out.sub$get_data.n1[i],".",
                              out.sub$get_data.n2[i]),
                           pval = fl$p,
                           stringsAsFactors = F)
  }
  names(res) <- as.character(out.sub$method)
  return(res)
}





plotTypeI <- function(res, alpha=.05,
                      title_label=NULL) {
  library(dplyr)
  res_summary <- do.call(rbind, res)
  res_summary <- res_summary %>% group_by(method) %>% summarise(type1=mean(pval<alpha))
  n_methods <- length(unique(res_summary$method))
  cols <- RColorBrewer::brewer.pal(n_methods,name="Accent")
  print(
    ggplot(res_summary, aes(x=method, y=type1, col=method)) +
      geom_point() +
    scale_color_manual(values=cols) )
}


plotPval <- function(res, seed=1) {
    library(dplyr)
    library(ggplot2)

    res_summary <- do.call(rbind, res)
    n_methods <- length(unique(res_summary$method))

    res_summary <- res_summary[res_summary$seed==seed,]
    
    cols <- RColorBrewer::brewer.pal(n_methods,name="Accent")
    res_summary$method <- as.factor(res_summary$method)
    print(
    ggplot(res_summary, aes(x=pval, fill=method)) +
            facet_wrap(~method) +
            geom_histogram() +
            scale_fill_manual(values=cols)  )
    }


plotQQ <- function(res, seed=1, sample_size=50*2, title_label=NULL,
                    mfrow_dim=c(2,2)) {
    res_summary <- do.call(rbind, res)
    n_methods <- length(unique(res_summary$method))
    res_summary <- res_summary[res_summary$seed==seed,]

    cols <- RColorBrewer::brewer.pal(n_methods,name="Accent")

    qq <- lapply(1:length(res), function(i) {
      qqplot(x=runif(sample_size*2,0,1), y=res[[i]]$pval, plot.it=F)
    })
    plot(qq[[1]]$x, qq[[1]]$y, col = cols[1], cex=.7, pch = 16,
         xlab = "Uniform(0,1)", ylab = "Empirical distribution",
         main = "QQ-plot")
    for (i in 2:n_methods) {
      points(qq[[i]]$x, qq[[i]]$y, col = cols[i], cex=.7, pch = 16)
    }
    abline(0,1, col = "black")
    title(title_label, outer=TRUE, line=-1)
}
```




---

## Results

* 50 per group, correlated samples 

```{r}
library(dscrutils)
#setwd("~/Dropbox/GitHub/dsc-log-fold-change/dsc")
dir_dsc <- "/scratch/midway2/joycehsiao/dsc-log-fold-change/benchmark"
out <- dscquery(dir_dsc, 
                c("get_data", "get_data.seed",
                  "get_data.n1", "get_data.n2", "method", "method.p"))

res_corr <- get_res(dir_dsc, out, sim_case="sample_correlated", sample_size=50)
plotTypeI(res_corr, alpha=.05)
plotPval(res_corr)
plotQQ(res, seed=1, sample_size=50*2)
```

* 300 per group, correlated samples

```{r}
res_corr <- get_res(dir_dsc, out, sim_case="sample_correlated", sample_size=300)
plotTypeI(res_corr, alpha=.05)
plotPval(res_corr)
plotQQ(res, seed=1, sample_size=300*2)
```


