---
title: "Effect of data transformation in null data"
author: "Joyce Hsiao"
date: "2019-05-07"
output: 
  workflowr::wflow_html:
      toc: true
      float: true
---

## Introduction

Applying data transformation methods to single-cell gene expression count and computing type I error rate in null datasets. 

---

Notations:

$X_{ig}$: gene expression count in cell $i$ from gene $g$

$X_{i+}$: gene expression count in cell $i$, i.e., $X_{i+} = \sum_g X_{ig}$

$S_i$: library size normalization factor for cell $i$

$p$: a positive pseudo-count added to expression matrix; traditionally it is used to ensure log-transformation of the expression matrix is well-defined. For now we use pseudo-count of 1.

<!-- Moreover, we assume that $X_{ig} \sim Pois(\lambda_{ig} S_i)$. -->

---

Data transformation compared:

1. `log2_none_p1`: $log2(X_{ig} + p)$ where $p=1$.

2. `log2_libsum_p1`: $log2(X_{ig}/X_{i+} + p)$ where $p=1$

3. `log2_libscale_TMM_p1`: $log2(X_{ig}/S_{i} + p)$ where $p=1$ using edgeR TMM method to estimate $S_i$.

4. `log2_libscale_RLE_p1`: $log2(X_{ig}/S_{i} + p)$ where $p=1$ using DESeq2 RLE method to estimate $S_i$.

5. `counts_pearsons`: Pearson's residuals of expression counts, derived using `sctransform` (Hafemeister and Satija, 2019). 

---

Pipeline compared:

* t-test + all 5 data transformation methods
* DEseq2: count data + RLE transformation
* edgeR: count data + TMM transformation

---

More about `counts_pearson`:

For a given gene $g$, use the sum of all molecules assigned to a cell as a proxy for sequencing depth, and use this cell attribute in a regression model with negative binomial distribution and log link function. Thus, let $X_g$ be the vector of UMI counts assigned to gene $g$, and $m$ be the vector of molecules assigned to the cells, i.e., $m_i = \sum_i X_{ig}$. For a given $g$, we have

$log(E(X_g)) = \beta_0 + \beta_1 log10 m$

Using the NB parametrization with mean $\mu$ and variance $\mu + \mu^2/\theta$,

Pearson's residuals are defined as:

$z_{ig} = (X_{ig}-\mu_{ig})/\sigma_{ig}$    

where 

$\mu_{ig} = exp(\beta_{0g} + \beta_{1g}log10 m_i)$,    

$\sigma_{ig} = \sqrt(\mu_{ig} + \mu^2_{ig}/\theta_{g})$

---

Data simulation parameters:

* 100 cells (50 vs 50)
* 1K genes randomly drawn
* 100% null genes
* 20 simulated datasets


---

## Required packages

```{r}
knitr::opts_chunk$set(warning=F, message=F)

library(dscrutils)
library(tidyverse)
```


## Simulate data and run methods

```{r, warning=F, message=F, eval=F}

#methods <- c("limma_voom", "sva_ttest", "t_test")

library(seqgendiff)
library(sctransform)
#source("dsc/modules/poisthin.R")
source("dsc/modules/filter_genes.R")
source("dsc/modules/transform_data.R")
source("dsc/modules/t_test.R")
source("dsc/modules/wilcoxon.R")
source("dsc/modules/limma_voom.R")
source("dsc/modules/edger.R")
source("dsc/modules/deseq2.R")

counts <- readRDS("dsc/data/pbmc_counts.rds")
nsamp <- 100
ngene <- 1000
prop_null <- 0
libsize_factor <- 0
signal_fun <- function(n) rep(libsize_factor, n)
signal_params <- list()
#pvals_thres <- .001
nsim <- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj <- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0("output/transform_null.Rmd/data_obj_",i,".rds"))
}


nsim <- 50    
transform_methods_list <- c("log2_none_p1", "log2_libsum_p1", "log2_libscale_TMM_p1", 
                            "log2_libscale_RLE_p1", "counts_pearsons")
de_methods_list <- c("edger", "deseq2", "limma_voom", "t_test")

transform_methods_list <- c("log2_none_p1")
de_methods_list <- c("t_test")

out <- do.call(rbind, lapply(1:nsim, function(i) {

  data_obj <- readRDS(file = paste0("output/transform_null.Rmd/data_obj_",i,".rds"))
  Y <- t(data_obj$Y)
  X <- data_obj$X
  keep_genes <- filter_genes(Y, min_cell_detected=5)
  Y <- Y[keep_genes,]
  
  foo_m <- do.call(rbind, lapply(1:length(de_methods_list), function(j) {

  if (de_methods_list[j] == "edger") {
      res <- edger(Y=Y, X=X)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_005=mean(pvals < .005, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = i))
  } 
  if (de_methods_list[j] == "deseq2") {
      res <- deseq2(Y=Y, X=X)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_005=mean(pvals < .005, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = i))
  } 
  if (de_methods_list[j] == "limma_voom") {
      res <- limma_voom(Y=Y, X=X)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_005=mean(pvals < .005, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = i))
  } 

  if (de_methods_list[j] == "t_test") {
      foo_t <- do.call(rbind, lapply(1:length(transform_methods_list), function(k) {
          if (transform_methods_list[k] == "log2_none_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "none", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libsum_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "sum", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_TMM_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "TMM", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_RLE_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "RLE", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "counts_pearsons") {
          transformed_Y <- transform_data(Y, libscale_method = "pearsons_residual", 
                                          log="none", pseudo_count=1)
          }
          res <- t_test(transformed_Y, X)
          pvals <- res[2,]
          return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                            type1error_005=mean(pvals < .005, na.rm=TRUE),
                            type1error_001=mean(pvals < .001, na.rm=TRUE),
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = i))
      }) )
      return(foo_t)      
    }
  }) )
   return(foo_m)
}))


saveRDS(out, file = "output/transform_null.Rmd/type1error.rds")
```



## Analysis


```{r, eval=T}
#alpha <- .001
out <- readRDS(file = "output/transform_null.Rmd/type1error.rds")

levels(out$transform_method)
labels_methods <- c("edger", "deseq2", "limma_voom", "ttest_log2_none_p1",
            "ttest_log2_libsum_p1", "ttest_log2_libscale_TMM_p1", 
            "ttest_log2_libscale_RLE_p1", "ttest_counts_pearsons")


out %>% #filter(n1==50) %>% 
    group_by(de_method, transform_method) %>%
    ggplot(., aes(x=transform_method, y=type1error_001, col=transform_method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .001") +
        geom_hline(yintercept = .001, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top",
                       labels=labels_methods) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
  
out %>% #filter(n1==50) %>% 
    group_by(de_method, transform_method) %>%
    ggplot(., aes(x=transform_method, y=type1error_01, col=transform_method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .01") +
        geom_hline(yintercept = .01, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top",
                       labels=labels_methods) +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")

```


