---
title: "Effect of data transformation in null data"
author: "Joyce Hsiao"
date: "2019-05-07"
output: 
  workflowr::wflow_html:
      toc: false
      float: true
---

## Summary

Applying data transformation methods to single-cell gene expression count and computing type I error rate in null datasets. 

* [When library sizes are the same between two groups](#no-differences-in-library-sizes), except for edgeR which has poor control over type I error at alpha < .01, all other methods are able to control for type I error. 

* [When sample library size in one group is 2 times the library size in the other group](#library-size-factor-1) all methods perform pooly, except for the method that applies t-test to Pearson's residuals (Hafemeister and Satija, 2019). We observed [similar results after increasing library size factor to 2](#library-size-factor-2) (group 1 sample size $2^2$ times of group 2 sample size).

---

Data transformation methods compared:

$X_{ig}$: gene expression count in cell $i$ from gene $g$

$X_{i+}$: gene expression count in cell $i$, i.e., $X_{i+} = \sum_g X_{ig}$

$S_i$: library size normalization factor for cell $i$

$p$: a positive pseudo-count added to expression matrix; traditionally it is used to ensure log-transformation of the expression matrix is well-defined. For now we use pseudo-count of 1.

1. `log2_none_p1`: $log2(X_{ig} + p)$ where $p=1$.

2. `log2_libsum_p1`: $log2(X_{ig}/X_{i+} + p)$ where $p=1$

3. `log2_libscale_TMM_p1`: $log2(X_{ig}/S_{i} + p)$ where $p=1$ using edgeR TMM method to estimate $S_i$.

4. `log2_libscale_RLE_p1`: $log2(X_{ig}/S_{i} + p)$ where $p=1$ using DESeq2 RLE method to estimate $S_i$.

5. `counts_pearsons`: Pearson's residuals of expression counts, derived using `sctransform` (Hafemeister and Satija, 2019). 

---

Pipelines compared:

* DEseq2: count data + RLE transformation
* edgeR: count data + TMM transformation
* limma_voom + data transformation 1-4
* t-test + all 5 data transformation methods listed above

---

More about `counts_pearson`:

For a given gene $g$, use the sum of all molecules assigned to a cell as a proxy for sequencing depth, and use this cell attribute in a regression model with negative binomial distribution and log link function. Thus, let $X_g$ be the vector of UMI counts assigned to gene $g$, and $m$ be the vector of molecules assigned to the cells, i.e., $m_i = \sum_i X_{ig}$. For a given $g$, we have

$log(E(X_g)) = \beta_0 + \beta_1 log10 m$

Using the NB parametrization with mean $\mu$ and variance $\mu + \mu^2/\theta$,

Pearson's residuals are defined as:

$z_{ig} = (X_{ig}-\mu_{ig})/\sigma_{ig}$    

where 

$\mu_{ig} = exp(\beta_{0g} + \beta_{1g}log10 m_i)$,    

$\sigma_{ig} = \sqrt(\mu_{ig} + \mu^2_{ig}/\theta_{g})$

---

Data simulation parameters:

* 100 cells (50 vs 50)
* 1K genes randomly drawn
* 100% null genes
* 50 simulated datasets


---

## Required packages and functions

```{r load_packages, warning=F, message=F, eval=T}

library(tidyverse)
library(seqgendiff)
library(sctransform)
#source("dsc/modules/poisthin.R")
source("dsc/modules/filter_genes.R")
source("dsc/modules/transform_data.R")
source("dsc/modules/t_test.R")
source("dsc/modules/wilcoxon.R")
source("dsc/modules/limma_voom.R")
source("dsc/modules/edger.R")
source("dsc/modules/deseq2.R")
source("dsc/modules/sva.R")

run_methods <- function(data_obj, nsim, verbose=F) {
  
  Y <- t(data_obj$Y)
  X <- data_obj$X
  beta <- data_obj$beta
  keep_genes <- filter_genes(Y, min_cell_detected=5)
  Y <- Y[keep_genes,]
  beta <- beta[keep_genes]
  
  foo_m <- do.call(rbind, lapply(1:length(de_methods_list), function(j) {

  if (de_methods_list[j] == "edger") {
    if (verbose) message("edger")
      res <- edger(Y=Y, X=X)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        mse = mean((res$est - beta)^2, na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == "deseq2") {
      if (verbose) message("deseq2")
      res <- deseq2(Y=Y, X=X)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        mse = mean((res$est - beta)^2, na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == "limma") {
      if (verbose) message("limma")
      foo_l <- do.call(rbind, lapply(1:(length(transform_methods_list)-1), function(k) {
          if (transform_methods_list[k] == "log2_none_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "none", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libsum_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "sum", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_TMM_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "TMM", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_RLE_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "RLE", 
                                          log="log2", pseudo_count=1)
          }
          if (sum(is.na(transformed_Y))==nrow(transformed_Y)*ncol(transformed_Y)) {
          return(data.frame(type1error_01=NA,
                            type1error_001=NA,
                            mse=NA,
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
          } else {
          fit <- lmFit(transformed_Y,design=X)
          fit.ebayes <- eBayes(fit)
          pvals <- fit.ebayes$p.value[,2]
          return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                            type1error_001=mean(pvals < .001, na.rm=TRUE),
                            mse=mean((fit.ebayes$coefficients[,2]-beta)^2,na.rm=T),
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
          }
      }) )
      return(foo_l)      
    }
  if (de_methods_list[j] == "limma_voom") {
      if (verbose) message("limma_voom")
      res <- limma_voom(Y=Y, X=X)
      pvals <- res$pvalue
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        mse=mean((res$betahat-beta)^2,na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == "limma_voom_libscale_tmm") {
  if (verbose) message("limma_voom_libscale_tmm")
  libnorm_factors <- edgeR::calcNormFactors(Y, method="TMM")
  res <- limma_voom(Y, X, libnorm_factors=libnorm_factors)
  pvals <- res$pvalue
  return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                    type1error_001=mean(pvals < .001, na.rm=TRUE),
                    mse=mean((res$betahat-beta)^2,na.rm=T),
                    transform_method = de_methods_list[j],
                    de_method = de_methods_list[j],
                    nsim = nsim))
  } 
  if (de_methods_list[j] == "limma_voom_libscale_rle") {
    if (verbose) message("limma_voom_libscale_rle")
    libnorm_factors <- edgeR::calcNormFactors(Y, method="RLE")
    if (anyNA(libnorm_factors)) {
    return(data.frame(type1error_01=NA,
                      type1error_001=NA,
                      mse=NA,
                      transform_method = de_methods_list[j],
                      de_method = de_methods_list[j],
                      nsim = nsim))
    } else {
    res <- limma_voom(Y, X, libnorm_factors=libnorm_factors)
    pvals <- res$pvalue
    return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                      type1error_001=mean(pvals < .001, na.rm=TRUE),
                      mse=mean((res$betahat-beta)^2,na.rm=T),
                      transform_method = de_methods_list[j],
                      de_method = de_methods_list[j],
                      nsim = nsim))
    }
  } 
  if (de_methods_list[j] == "sva_limma_voom") {
      if (verbose) message("sva_limma_voom")
      output_sva <- sva(Y, X)
      res <- limma_voom(Y, X=output_sva$X.sv)
      pvals <- res$pval
      return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                        type1error_001=mean(pvals < .001, na.rm=TRUE),
                        mse=mean((res$betahat-beta)^2,na.rm=T),
                        transform_method = de_methods_list[j],
                        de_method = de_methods_list[j],
                        nsim = nsim))
  } 
  if (de_methods_list[j] == "t_test") {
      if (verbose) message("t_test")
      foo_t <- do.call(rbind, lapply(1:length(transform_methods_list), function(k) {
          if (transform_methods_list[k] == "log2_none_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "none", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libsum_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "sum", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_TMM_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "TMM", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "log2_libscale_RLE_p1") {
          transformed_Y <- transform_data(Y, libscale_method = "RLE", 
                                          log="log2", pseudo_count=1)
          }
          if (transform_methods_list[k] == "counts_pearsons") {
          transformed_Y <- transform_data(Y, libscale_method = "pearsons_residual", 
                                          log="none", pseudo_count=1)
          }
          res <- t_test(transformed_Y, X)
          pvals <- res[2,]
          return(data.frame(type1error_01=mean(pvals < .01, na.rm=TRUE),
                            type1error_001=mean(pvals < .001, na.rm=TRUE),
                            mse=mean((res[1,]-beta)^2,na.rm=T),
                            transform_method = transform_methods_list[k],
                            de_method = de_methods_list[j],
                            nsim = nsim))
      }) )
      return(foo_t)      
    }
  }) )
   return(foo_m)
}


transform_methods_list <- c("log2_none_p1", "log2_libsum_p1", "log2_libscale_TMM_p1", 
                            "log2_libscale_RLE_p1", "counts_pearsons")
de_methods_list <- c("edger", "deseq2", "limma", "limma_voom", 
                     "limma_voom_libscale_tmm",
                     "limma_voom_libscale_rle",
                     "sva_limma_voom", "t_test")
```


---

## No differences in library sizes 

Simulations

```{r simulations_null, warning=F, message=F, eval=F}
counts <- readRDS("dsc/data/pbmc_counts.rds")
nsamp <- 100
ngene <- 1000
prop_null <- 0
libsize_factor <- 0
signal_fun <- function(n) rep(libsize_factor, n)
signal_params <- list()
#pvals_thres <- .001
nsim <- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj <- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0("output/transform_null.Rmd/data_obj_",i,".rds"))
}



out <- do.call(rbind, lapply(1:nsim, function(i) {
  print(i)
  data_obj <- readRDS(file = paste0("output/transform_null.Rmd/data_obj_",i,".rds"))

  res <- run_methods(data_obj, i, verbose=T)
  return(res)
  }) )

out$method <- as.character(out$de_method)
which_relabel <- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] <- paste(as.character(out$de_method), as.character(out$transform_method), sep="_")[which_relabel]

out$method <- factor(out$method)

saveRDS(out, file = "output/transform_null.Rmd/libfactor_0.rds")
```

Results

```{r results, eval=T, fig.width=8}
#alpha <- .001
out <- readRDS(file = "output/transform_null.Rmd/libfactor_0.rds")
levels(out$method)


out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .001") +
        geom_hline(yintercept = .001, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
  
out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .01") +
        geom_hline(yintercept = .01, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")

out %>% #filter(n1==50) %>% 
    group_by(nsim, method) %>%
    summarise(mse_mn =mean(mse, na.rm=T)) %>%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Mean squared error") +
        ylab("Mean squared error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
```


---


## library size factor 1

Simulation

```{r simulate_data_libfactor_1, warning=F, message=F, eval=F}
counts <- readRDS("dsc/data/pbmc_counts.rds")
nsamp <- 100
ngene <- 1000
prop_null <- 0
libsize_factor <- 1
signal_fun <- function(n) rep(libsize_factor, n)
signal_params <- list()
#pvals_thres <- .001
nsim <- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj <- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0("output/transform_null.Rmd/data_obj_libfactor_1_",i,".rds"))
}


nsim=50
out <- do.call(rbind, lapply(1:nsim, function(i) {

  data_obj <- readRDS(file = paste0("output/transform_null.Rmd/data_obj_libfactor_1_",
                                    i,".rds"))

  res <- run_methods(data_obj, i)
  return(res)
  }) )

out$method <- as.character(out$de_method)
which_relabel <- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] <- paste(as.character(out$de_method), as.character(out$transform_method), sep="_")[which_relabel]

out$method <- factor(out$method)
saveRDS(out, file = "output/transform_null.Rmd/libfactor_1.rds")
```

Results

```{r results_libfactor_1, eval=T, fig.width=8}
#alpha <- .001
out <- readRDS(file = "output/transform_null.Rmd/libfactor_1.rds")

levels(out$method)

out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .001") +
        geom_hline(yintercept = .001, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
  
out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .01") +
        geom_hline(yintercept = .01, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")

out %>% #filter(n1==50) %>% 
    group_by(nsim, method) %>%
    summarise(mse_mn =mean(mse, na.rm=T)) %>%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Mean squared error") +
        ylab("Mean squared error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
```




---

## library size factor 2 


```{r simulate_data_libfactor_2, warning=F, message=F, eval=F}
counts <- readRDS("dsc/data/pbmc_counts.rds")
nsamp <- 100
ngene <- 1000
prop_null <- 0
libsize_factor <- 2
signal_fun <- function(n) rep(libsize_factor, n)
signal_params <- list()
#pvals_thres <- .001
nsim <- 50
for (i in 1:nsim) {
  set.seed(i)
  data_obj <- poisthin(t(counts), nsamp=nsamp, ngene=ngene, 
                       signal_params=signal_params, signal_fun=signal_fun, 
                       prop_null = prop_null)
  saveRDS(data_obj, file = paste0("output/transform_null.Rmd/data_obj_libfactor_2_",i,".rds"))
}


nsim=50
out <- do.call(rbind, lapply(1:nsim, function(i) {
  print(i)
  data_obj <- readRDS(file = paste0("output/transform_null.Rmd/data_obj_libfactor_2_",
                                    i,".rds"))

  res <- run_methods(data_obj, i, verbose = T)
  return(res)
  }) )
out$method <- as.character(out$de_method)
which_relabel <- which(as.character(out$de_method) != as.character(out$transform_method))
out$method[which_relabel] <- paste(as.character(out$de_method), as.character(out$transform_method), sep="_")[which_relabel]
out$method <- factor(out$method)

saveRDS(out, file = "output/transform_null.Rmd/libfactor_2.rds")
```

Results

```{r results_libfactor_2, eval=T, fig.width=8}
#alpha <- .001
out <- readRDS(file = "output/transform_null.Rmd/libfactor_2.rds")

levels(out$method)

out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_001, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .001") +
        geom_hline(yintercept = .001, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
  
out %>% #filter(n1==50) %>% 
    group_by(method) %>%
    ggplot(., aes(x=method, y=type1error_01, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .01") +
        geom_hline(yintercept = .01, col="gray50") +
        ylab("Type I error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")

out %>% #filter(n1==50) %>% 
    group_by(nsim, method) %>%
    summarise(mse_mn =mean(mse, na.rm=T)) %>%
    ggplot(., aes(x=method, y=mse_mn, col=method)) +
#        facet_wrap(~de_method) +
        geom_boxplot() + geom_point() + xlab("Mean squared error") +
        ylab("Mean squared error") +
      scale_x_discrete(position = "top") +
       theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
       stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
```


---

TBD

```{r, eval=F}
res_libfactor_null <- readRDS(file = "output/transform_null.Rmd/type1error.rds")
res_libfactor_null$libfactor <- 2^0
res_libfactor_1 <- readRDS(file = "output/transform_null.Rmd/type1error_libfactor_1.rds")
res_libfactor_1$libfactor <- 2^1
res_libfactor_2 <- readRDS(file = "output/transform_null.Rmd/type1error_libfactor_2.rds")
res_libfactor_2$libfactor <- 2^2

res <- rbind(res_libfactor_null,
             res_libfactor_1)
res$libfactor <- factor(res$libfactor)


# res %>% #filter(n1==50) %>% 
#     group_by(libfactor, transform_method) %>%
#     ggplot(., aes(x=transform_method, y=type1error_001, col=transform_method)) +
# #        facet_wrap(~libfactor) +
#         geom_boxplot() + geom_point() + xlab("Type 1 error at alpha < .001") +
#         geom_hline(yintercept = .001, col="gray50") +
#         ylab("Type I error") +
#       scale_x_discrete(position = "top",
#                        labels=labels_methods) +
#        theme(axis.text.x=element_text(angle = 20, vjust = -.3, hjust=.1)) +
#        stat_summary(fun.y=mean, geom="point", shape=4, size=4, col="black")
  

```

